##file: script_config.R

###########################################################################
## Automated Options: DON'T EDIT BELOW HERE UNLESS YOU HAVE A REASON
###########################################################################
##path to trimmed, merged fastq files
##trim and merge using trim_fastq.R
trimmed_dir <- paste(output_dir, "trimmed", sep="/")

##path to your fasta file directory
##make sure there are no other fasta files in this directory
fasta_dir <- paste(output_dir, "fasta_files", sep="/")

##paths to your merged, cleaned final.fa file
final.fa <- paste(output_dir, "merged/final.fa", sep="/")

##path to merged,trimmed forward read file
left <- paste(output_dir, "trimmed/trimmed_left.fq", sep="/")

##path to merged,trimmed reverse read file
right <- paste(output_dir, "trimmed/trimmed_right.fq", sep="/")

##path to merged single fq file
single <- paste(output_dir, "trimmed/trimmed_single.fq", sep="/")

##full path to RSEM.genes.results file generated by running RSEM
gene_FPKM <-paste(output_dir, "base_files/RSEM.genes.results", sep="/")

##full path to RSEM.isoforms.results file generated by running RSEM
isoforms_FPKM <-paste(output_dir, "base_files/RSEM.isoforms.results", sep="/")

##full path to the tab separated blastx report for this fasta
blastx_report <- paste(output_dir, "base_files/final.fa_blastx", sep="/")
blastx_chick <- paste(output_dir, "base_files/final.fa_blastx_chick", sep="/")
blastp_report <- paste(output_dir, "peptide/final.fa_blastp", sep="/")

##path to csorted bam file for full assembly
bam_file <- paste(output_dir, "base_files/bowtie.csorted.bam", sep="/")

##enter the path to the filtered fasta file created with filter_fasta.R
filtered.fa <- paste(output_dir, "filtering/final_merged.fa", sep="/")

##path to the filtered blastx report
filtered_blastx <- paste(output_dir, "/filtering/final_merged.fa_blastx", sep="/")

##path to RSEM generated genes_to_transcripts mapping file
genes_to_trans <- paste(output_dir, "/base_files/RSEM.isoforms.results", sep="")

